splice.nix: inherit from lib

Artturin opened the pull request:
This may improve performance

For stdenv inheriting from lib improved performance c3c31aa798551a2808981d14537d04de8f51dff8

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Artturin committed:
splice.nix: inherit from `lib`

This may improve performance

For stdenv inheriting from lib improved performance c3c31aa798551a2808981d14537d04de8f51dff8

philiptaron commented on pkgs/top-level/splice.nix:
```diff
@@ -102,7 +116,7 @@ let
     , pkgsTargetTarget
     } @ args:
-    if actuallySplice then spliceReal args else pkgsHostTarget;
+    if true then spliceReal args else pkgsHostTarget;
```
This isn't intentional, right?

Artturin replied:
It's so ofborg uses the splicing path

philiptaron replied:
Makes total sense. I'm very interested in this (and similar) PRs!

philiptaron commented on pkgs/top-level/splice.nix:
```diff
@@ -17,6 +17,20 @@
 lib: pkgs: actuallySplice:
 
 let
+  inherit (lib)
+    attrByPath
+    callPackageWith
+    callPackagesWith
+    genAttrs
+    isAttrs
+    isDerivation
+    listToAttrs
+    makeScopeWithSplicing
+    makeScopeWithSplicing'
+    optional
+    optionalAttrs
+    splitString
+    ;
```
You should be able to hoist this between the `lib:` and `pkgs:` so it's shared over a wider context, too.

Artturin committed:
Try

Artturin commented:
Eval failed and of course it's QT.... https://gist.github.com/GrahamcOfBorg/f132dfbaac921f6172f0728c8b4c13df

Artturin commented:
Had to stop returning of cross/always splice because they've been killing ofborg lol

Artturin commented:
@ofborg eval

Run 1 
![image](https://github.com/user-attachments/assets/c93ab56e-f83e-4e7d-8d05-adfda7cc915c)

kjeremy commented:
Looks pretty neutral

Artturin committed:
remove heavy python override stuff to allow ofborg to eval

Artturin committed:
Return cross pkgs to test perf

kjeremy commented:
I think the remaining eval errors are resolved

philiptaron commented:
I am loving watching this progress through the thicket of eval errors!

philiptaron commented:
Hell yeah, a successful eval!

Artturin commented:
> Hell yeah, a successful eval!

Yeah now gotta wait for a ofborg eval, github actions eval doesn't have perf stats.

philiptaron commented:
I'm sort of assuming that you'll make separate PRs for each of these DONTDROP commits; they all look worth taking in some shape or form.

Artturin commented:
> I'm sort of assuming that you'll make separate PRs for each of these DONTDROP commits; they all look worth taking in some shape or form.

https://github.com/NixOS/nixpkgs/pull/362546

philiptaron commented:
Dang, ofBorg barfed on a log line.

https://gist.githubusercontent.com/GrahamcOfBorg/8546e456a1cef3658055c48d37902deb/raw/841f915c5b3e68493a309edee7d279edcdb9bdc7/Output%2520path%2520comparison

Artturin commented:
On dfd6a4a50f9021fcda82f5151642e5e13670d4f2 the eval succeeds, after rebasing to master the eval doesn't succeed :/

Artturin commented:
ofborg-eval-lib-tests fail

https://gist.github.com/GrahamcOfBorg/a2b431acc304bc74cdb3723cb560980f

`       > /build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 15: /nix/store/7pami0aqan88pykzki6rqswmm54kv9ir-nix-aarch64-unknown-linux-gnu-2.24.10/bin/nix-instantiate: cannot execute binary file: Exec format error`

Fixed

Artturin committed:
nixosOptionsDoc: Fix cross build

Artturin committed:
Return cross pkgs only on x86_64-linux or aarch64-linux for testing

because of https://gist.github.com/GrahamcOfBorg/aa78dac437d7e9d77fb47cd8142615c7

philiptaron commented:
https://gist.github.com/GrahamcOfBorg/458aadffab9d4c8d51dc191cff93c222

Seems to hit every branch on the way down!

Artturin commented:
That was fixed in https://github.com/NixOS/nixpkgs/pull/364542

philiptaron commented:
https://gist.githubusercontent.com/GrahamcOfBorg/6627bea2a8da597c22b94008d04f29e2/raw/9fc4b15b83ebdfe7e163ecb1de33b3f9144e18c5/ofborg-eval-nixpkgs-unstable-jobset

qemu failure this time about darwin.

Artturin committed:
Hopefully this fixes the error

https://gist.github.com/GrahamcOfBorg/6627bea2a8da597c22b94008d04f29e2

>        error: Unsupported guest system aarch64-linux for host x86_64-darwin, supported: x86_64-linux

philiptaron commented:
https://gist.githubusercontent.com/GrahamcOfBorg/ae93ea34c749ca2928e67324f8275b83/raw/cbf13286041a1c167ed81c360b15e4cd6bf3a3d6/ofborg-eval-darwin

darwin GHC assert this time

kjeremy commented:
@Artturin Can we see what this looks like now that python cross has been improved?
